Source/WTF/wtf/glib/SysprofAnnotator.h

```diff
@@ -0,0 +1,456 @@
+/*
+ * Copyright (C) 2024 Igalia, S.L.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Library General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include <sysprof-capture.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/Vector.h>
+#include <wtf/text/ASCIILiteral.h>
+
+namespace WTF {
+
+class SysprofAnnotator final {
+    WTF_MAKE_NONCOPYABLE(SysprofAnnotator);
+
+    using RawPointerPair = std::pair<const void*, const void*>;
+    using TimestampAndString = std::pair<int64_t, Vector<char>>;
+
+public:
+
+    static SysprofAnnotator* singletonIfCreated()
+    {
+        return s_annotator;
+    }
+
+    void instantMark(std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, 0, m_processName, name.data(), description, args);
+        va_end(args);
+    }
+
+    void mark(Seconds timeDelta, std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, timeDelta.microsecondsAs<int64_t>(), m_processName, name.data(), description, args);
+        va_end(args);
+    }
+
+    void beginMark(const void* pointer, std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        auto key = std::make_pair(pointer, static_cast<const void*>(name.data()));
+
+        Vector<char, 1024> buffer;
+        va_list args;
+        va_start(args, description);
+        vsnprintf(buffer.data(), 1024, description, args);
+        va_end(args);
+
+        auto value = std::make_pair(SYSPROF_CAPTURE_CURRENT_TIME, WTFMove(buffer));
+
+        Locker locker { m_lock };
+        m_ongoingMarks.set(key, value);
```

Comment:
Does this actually save any chars in the map? isn't the `size()` of `buffer` 0? So when it gets copied to the HashMap it will be empty, IIUC? I saw this when refactoring for https://github.com/WebKit/WebKit/pull/29744 but I just fixed the build error I didn't change the semantics here.

Reply:
I wondered the same when I reviewed this, and IIRC size it's not 0, because of the stack allocated size 1024.

Reply:
I'm not certain, but I'd guess capacity is 1024 but size is actually 0. Probably not good even if it works. I think we need to verify this.

Reply:
Yes, but we are not using `size()`, we pass 1024 directly to `vsnprintf()`. I agree we should probably resize the buffer to be sure. I guess we could just change the declaration to:

```
Vector<char, 1024> buffer(1024);
```

Reply:
@GeorgesStavracas note there is action required here

> Vector<char, 1024> buffer(1024);

I bet we don't need to specify the capacity at all if we construct it with a size.

Reply:
> @GeorgesStavracas note there is action required here
> 
> > Vector<char, 1024> buffer(1024);
> 
> I bet we don't need to specify the capacity at all if we construct it with a size.

It does make a difference though, in one case, the bytes will be stack allocated, in the other heap allocated.

Reply:
OK, that's pretty important, and I didn't know.

So this Vector is going to be moved into a hash table with a lifetime that extends beyond the current stack frame. I assume the data will be copied (and not left dangling) if using the inline capacity. So not using inline capacity seems better here, yes?

Reply:
https://github.com/WebKit/WebKit/pull/31816